internal/controller/nats/controller.go

```diff
@@ -82,23 +86,87 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	// copy the object, so we don't modify the source object
 	nats := currentNats.DeepCopy()
 
+	// logger with nats details
+	log := r.loggerWithNATS(nats)
+
 	// check if nats is in deletion state
 	if nats.IsInDeletion() {
-		return ctrl.Result{}, nil
+		return r.handleNATSDeletion(ctx, nats, log)
 	}
 
 	// handle reconciliation
-	return ctrl.Result{}, nil
+	return r.handleNATSReconcile(ctx, nats, log)
+}
+
+// generateNatsResources renders the NATS chart with provided overrides.
+// It puts results into ReleaseInstance.
+func (r *Reconciler) generateNatsResources(nats *natsv1alpha1.NATS, instance *chart.ReleaseInstance) error {
+	// generate Nats resources from chart
+	natsResources, err := r.NATSManager.GenerateNATSResources(
+		instance,
+		manager.WithOwnerReference(*nats), // add owner references to all resources
+	)
+	if err != nil {
+		return err
+	}
+
+	// update manifests in instance
+	instance.SetRenderedManifests(*natsResources)
+	return nil
+}
+
+// initNATSInstance initializes a new NATS release instance based on NATS CR.
+func (r *Reconciler) initNATSInstance(ctx context.Context, nats *natsv1alpha1.NATS,
+	log *zap.SugaredLogger) (*chart.ReleaseInstance, error) {
+	// Init a release instance
+	instance := &chart.ReleaseInstance{
+		Name:      nats.Name,
+		Namespace: nats.Namespace,
+	}
+
+	// Check if istio is enabled in cluster
+	istioExists, err := r.kubeClient.DestinationRuleCRDExists(ctx)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Istio enabled on cluster: %t", istioExists)
+
+	accountSecretName := fmt.Sprintf("%s-secret", nats.Name)
+	// Check if secret exists then make sure the password is same
+	accountSecret, err := r.kubeClient.GetSecret(ctx, accountSecretName, nats.Namespace)
+	if err != nil && !errors.IsNotFound(err) {
+		log.Errorf("Failed to fetch secret: %s", accountSecretName)
+		log.Error(err)
+		return nil, err
+	}
+	log.Infof("NATS account secret (name: %s) exists: %t", accountSecretName, accountSecret != nil)
+
+	// @TODO: Provide the overrides in component.Configuration
+	instance.Configuration = map[string]interface{}{
+		"istio.enabled":       istioExists,
+		"auth.rotatePassword": accountSecret == nil, // do not recreate secret if it exists
```

Comment:
Are we sure this will create a secret that matches the `accountSecretName` and `nats.Namespace`?

Reply:
The name format is the same in NATS helm chart. If someone changes the name format in helm, the pre-main tests should fail.